File: lib/store.js
const fs = require("fs");
const path = require("path");

const SID = /^[A-Za-z0-9_-]{1,64}$/;

function isValidSid(sid) {
  return !!sid && SID.test(sid);
}

function loadFavoriteOrder(p) {
  try { const a = JSON.parse(fs.readFileSync(p, "utf-8")); return Array.isArray(a) ? a : []; }
  catch (e) { return []; }
}
function loadFavorites(p) {
  return new Set(loadFavoriteOrder(p));
}
function saveFavoriteOrder(p, arr) {
  fs.writeFileSync(p, JSON.stringify(arr));
}
function toggleFavorite(p, sid) {
  const a = loadFavoriteOrder(p);
  const i = a.indexOf(sid);
  let added;
  if (i >= 0) { a.splice(i, 1); added = false; } else { a.push(sid); added = true; }
  saveFavoriteOrder(p, a);
  return added;
}
function moveFavorite(p, sid, dir) {
  const a = loadFavoriteOrder(p);
  const i = a.indexOf(sid);
  if (i < 0) return false;
  const j = i + (dir < 0 ? -1 : 1);
  if (j < 0 || j >= a.length) return false;
  const t = a[i]; a[i] = a[j]; a[j] = t;
  saveFavoriteOrder(p, a);
  return true;
}

function loadTrashMeta(p) {
  try { return JSON.parse(fs.readFileSync(p, "utf-8")); } catch (e) { return {}; }
}
function saveTrashMeta(p, d) {
  fs.writeFileSync(p, JSON.stringify(d));
}

function deleteSession(file, sid, trashDir, metaPath) {
  if (!isValidSid(sid)) return false;
  try { fs.mkdirSync(trashDir, { recursive: true }); } catch (e) {}
  const dest = path.join(trashDir, path.basename(sid) + ".jsonl");
  try {
    fs.renameSync(file, dest);
  } catch (e) {
    try { fs.copyFileSync(file, dest); fs.unlinkSync(file); } catch (e2) { return false; }
  }
  const meta = loadTrashMeta(metaPath);
  meta[sid] = { session_id: sid, origin: file, deleted_at: Date.now() / 1000 };
  saveTrashMeta(metaPath, meta);
  return true;
}

function restoreSession(sid, trashDir, metaPath) {
  if (!isValidSid(sid)) return false;
  const meta = loadTrashMeta(metaPath);
  const info = meta[sid];
  if (!info) return false;
  const projRoot = path.resolve(path.dirname(trashDir));
  const dest = path.resolve(info.origin);
  if (!(dest === projRoot || dest.startsWith(projRoot + path.sep))) return false;
  const src = path.join(trashDir, path.basename(sid) + ".jsonl");
  if (fs.existsSync(dest)) return false;
  try {
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    fs.renameSync(src, dest);
  } catch (e) { return false; }
  delete meta[sid];
  saveTrashMeta(metaPath, meta);
  return dest;
}

function listTrash(metaPath) {
  return Object.values(loadTrashMeta(metaPath)).sort(
    (a, b) => (b.deleted_at || 0) - (a.deleted_at || 0));
}

function renameSession(file, sid, title) {
  fs.appendFileSync(file,
    JSON.stringify({ type: "custom-title", customTitle: title, sessionId: sid }) + "\n", "utf-8");
  return true;
}

module.exports = { isValidSid, loadFavorites, loadFavoriteOrder, toggleFavorite, moveFavorite,
  deleteSession, restoreSession, listTrash, renameSession };

File: csm/store.py
import os, json, shutil, time, re

_SID_RE = re.compile(r"\A[A-Za-z0-9_-]{1,64}\Z")

def is_valid_sid(sid):
    return bool(sid) and bool(_SID_RE.match(sid))

def load_favorites(path):
    if not os.path.exists(path):
        return set()
    try:
        with open(path, encoding="utf-8") as f:
            return set(json.load(f))
    except Exception:
        return set()

def _save_favorites(path, favs):
    with open(path, "w", encoding="utf-8") as f:
        json.dump(sorted(favs), f, ensure_ascii=False)

def toggle_favorite(path, sid):
    favs = load_favorites(path)
    if sid in favs:
        favs.discard(sid)
        added = False
    else:
        favs.add(sid)
        added = True
    _save_favorites(path, favs)
    return added

def _load_trash_meta(path):
    if not os.path.exists(path):
        return {}
    try:
        with open(path, encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return {}

def _save_trash_meta(path, data):
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False)

def delete_session(file_path, sid, trash_dir, meta_path):
    if not is_valid_sid(sid):
        return False
    os.makedirs(trash_dir, exist_ok=True)
    dest = os.path.join(trash_dir, os.path.basename(sid) + ".jsonl")
    try:
        shutil.move(file_path, dest)
    except (OSError, PermissionError):
        return False
    meta = _load_trash_meta(meta_path)
    meta[sid] = {"session_id": sid, "origin": file_path, "deleted_at": time.time()}
    _save_trash_meta(meta_path, meta)
    return dest

def restore_session(sid, trash_dir, meta_path):
    if not is_valid_sid(sid):
        return False
    meta = _load_trash_meta(meta_path)
    info = meta.get(sid)
    if not info:
        return False
    proj_root = os.path.realpath(os.path.dirname(trash_dir))
    dest = os.path.realpath(info["origin"])
    try:
        if os.path.commonpath([proj_root, dest]) != proj_root:
            return False
    except ValueError:
        return False
    src = os.path.join(trash_dir, os.path.basename(sid) + ".jsonl")
    os.makedirs(os.path.dirname(dest), exist_ok=True)
    shutil.move(src, dest)
    del meta[sid]
    _save_trash_meta(meta_path, meta)
    return True

def list_trash(meta_path):
    return sorted(_load_trash_meta(meta_path).values(),
                  key=lambda x: x.get("deleted_at", 0), reverse=True)

def rename_session(file_path, sid, new_title):
    record = {"type": "custom-title", "customTitle": new_title, "sessionId": sid}
    with open(file_path, "a", encoding="utf-8") as f:
        f.write(json.dumps(record, ensure_ascii=False) + "\n")
    return True
